src/libraries/System.Text.Json/System.Text.Json.SourceGeneration/JsonSourceGenerator.cs

```diff
@@ -16,47 +18,105 @@ namespace System.Text.Json.SourceGeneration
-        }
-    }
-}");
+namespace HelloWorldGenerated
+{{
+    public class {entry.Key}ClassInfo
+    {{
+        public {entry.Key}ClassInfo() {{ }}
 
-            context.AddSource("helloWorldGenerated", SourceText.From(sourceBuilder.ToString(), Encoding.UTF8));
+        public string TestMethod()
+        {{
+            return ""{entry.Key}"";
+        }}
+    }}
+}}
+", Encoding.UTF8));
+            }
         }
 
         public void Initialize(InitializationContext context)
         {
             context.RegisterForSyntaxNotifications(() => new JsonSerializableSyntaxReceiver());
         }
 
-        // Temporary function for now that reads all types. Should search types with attribute JsonSerializable.
-        internal class JsonSerializableSyntaxReceiver : ISyntaxReceiver
+        public class JsonSerializableSyntaxReceiver : ISyntaxReceiver
         {
-            public List<TypeDeclarationSyntax> GeneratorInputTypes = new List<TypeDeclarationSyntax>();
+            public Dictionary<string, IdentifierNameSyntax> ExternalClassTypeDict = new Dictionary<string, IdentifierNameSyntax>();
+            public Dictionary<string, TypeDeclarationSyntax> InternalClassTypeDict = new Dictionary<string, TypeDeclarationSyntax>();
 
             public void OnVisitSyntaxNode(SyntaxNode syntaxNode)
             {
-                // Get all the type decl in all syntax tree.
-                if (syntaxNode is TypeDeclarationSyntax tds)
+                // Look for classes or structs for JsonSerializable Attribute.
+                if (syntaxNode is ClassDeclarationSyntax || syntaxNode is StructDeclarationSyntax)
                 {
-                    GeneratorInputTypes.Add(tds);
+                    // Find JsonSerializable Attributes.
+                    IEnumerable <AttributeSyntax> serializableAttributes = Enumerable.Empty<AttributeSyntax>();
+                    AttributeListSyntax attributeList = ((TypeDeclarationSyntax)syntaxNode).AttributeLists.SingleOrDefault();
+                    if (attributeList != null)
+                    {
+                        serializableAttributes = attributeList.Attributes.Where(node => (node is AttributeSyntax attr && attr.Name.ToString() == "JsonSerializable")).Cast<AttributeSyntax>();
+                    }
+                    if (serializableAttributes.Any())
+                    {
+                        foreach (AttributeSyntax attributeNode in serializableAttributes)
+                        {
+                            // Check if the attribute is being passed a type.
+                            if (attributeNode.DescendantNodes().Where(node => node is TypeOfExpressionSyntax).Any())
+                            {
+                                // Get JsonSerializable attribute arguments.
+                                AttributeArgumentSyntax attributeArgumentNode = (AttributeArgumentSyntax)attributeNode.DescendantNodes().Where(node => node is AttributeArgumentSyntax).SingleOrDefault();
```

Comment:
What happens if someone has a custom `JsonSerializableAttribute` with usage like this on a type `[JsonSerializable(string, [...], Type, [...])]`?